R/kmeans-clustering-method.R

```diff
@@ -0,0 +1,528 @@
+#' @examples
+#' 
+#' #### Example using iris dataset ####
+#' ptm <- proc.time()
+#' library(healthcareai)
+#' 
+#' data(iris)
+#' head(iris)
+#' 
+#' set.seed(2017)
+#' 
+#' p <- UnsupervisedModelParams$new()
+#' p$dataType <- "numeric"
+#' p$df <- iris
+#' p$labelCol <- "Species"
+#' p$impute <- TRUE
+#' p$debug <- FALSE
+#' p$cores <- 1
+#' 
+#' # Run k means clustering
+#' cl <- KmeansClustering$new(p)
+#' cl$run()
+#' 
+#' # Get the fit result
+#' cl$getKmeansfit()
+#' 
+#' # Get the elbow plot which also presents the optimal number of clusters
+#' cl$getElbowPlot()
+#' 
+#' # Get the 2D representation of the cluster solution
+#' cl$plotClusters()
+#' 
+#' # Get the sillhouette plot
+#' cl$silhouettePlot()
+#'  
+#' # Get a confusion matrix if labelCol exists
+#' cl$getConfusionMatrix()
+#' 
+#' # Get cluster labels
+#' cl$getClusterLabels() ## cluster 1 is labeled with "versicolor"
+#'                       ## cluster 2 is labeled with "setosa"
+#'                       ## cluster 3 is labeled with "virginica"
+#' 
+#' # Get the output data frame
+#' dfOut <- cl$getOutDf()
+#' head(dfOut)
+#' 
+#' ## Write to CSV (or JSON, MySQL, etc) using plain R syntax
+#' ## write.csv(dfOut,'path/predictionsfile.csv')
+#' 
+#' print(proc.time() - ptm)
+#' 
+#' #### Example using iris data and PCA ####
+#' ptm <- proc.time()
+#' library(healthcareai)
+#' 
+#' set.seed(2017)
+#' 
+#' p <- UnsupervisedModelParams$new()
+#' p$dataType <- "numeric"
+#' p$df <- iris
+#' p$labelCol <- "Species"
+#' p$cores <- 1
+#' p$pca <- TRUE ## set pca = TRUE
+#' 
+#' # Run k means clustering
+#' cl <- KmeansClustering$new(p)
+#' cl$run()
+#' 
+#' # Get the scree plot
+#' ## This plot presents the fraction of total variance in the data as explained
+#' ## or represented by each principle component.
+#' cl$getScreePlot()
+#' 
+#' # According to the scree plot, we decide to use the first 3 PCs to do clustering.
+#' p$usePrinComp <- TRUE
+#' p$numOfPrinComp <- 3
+#' 
+#' # Run k means clustering
+#' cl <- KmeansClustering$new(p)
+#' cl$run()
+#' 
+#' # Get the fit result
+#' cl$getKmeansfit()
+#'
+#' @export
+
+KmeansClustering <- R6Class("KmeansClustering",
+                            
+  # Inheritance
+  inherit = UnsupervisedModel,
+  
+  # Private members
+  private = list(
+    # Get kmeans model
+    kmeans.fit = NA,
+    confusionMatrix = NA,
+    vnum = NA,
+    prin.comp = NA,
+    df.prin = NA,
+    prop_varex = NA,
+    centers = NA,
+    cluster = NA,
+    outDf = NA,
+    cluster.labels = NA,
+    mean.vec = NA,
+    sd.vec = NA,
+    scaledf = NA,
+
+    
+    #check if the data type is numeric
+    checkDataType = function() {
+      #print(self$params$dataType)
+      if (self$params$dataType != 'numeric' || isNumeric(self$params$df) == FALSE) {
+        stop("Your data type must be numeric in order to use kmeans.")
+      }
+    },
+    
+    # Scale the data
+    dataScale = function() {
```

Comment:
Would you move `dataScale`, `pcaAnalysis`, and `distancePointSegment`, `findElbow`, `buildClusters`,`assignClusterLabels` to a new common-kmeans.R file and have unit tests around them? This file shouldn't really have complicated code, but rather pulls together functions that are well-documented in common-kmeans.R. As is, this code would be hard to debug. It's also nice that in common we'd have roxygen docs explaining the functions super clearly. I know it's clear to you now, but wait three months :)